pkg/px-resource-gateway/semaphore.go

```diff
@@ -0,0 +1,310 @@
+		priorityQ:           NewPriorityQueue(),
+		locks:               locks,
+		availableLocks:      availableLocks,
+		heartbeats:          heartbeats,
+		configMap:           configMap,
+		configMapUpdateDone: make(chan struct{}),
+	}
+
+	// start background workers
+	go semPQ.configMapUpdateRoutine()
+	go semPQ.cleanupDeadNodesRoutine()
+
+	return semPQ
+}
+
+func (s *semaphorePriorityQueue) AcquireLock(clientId string, priority pb.SemaphoreAccessPriority_Type) (pb.SemaphoreAccessStatus_Type, error) {
+	s.mutex.Lock()
+	updateDone := false
+	defer func() {
+		s.mutex.Unlock()
+		if updateDone {
+			<-s.configMapUpdateDone
+		}
+	}()
+	logrus.Infof("Received AcquireLock request for node %v", clientId)
+
+	_, alreadyExists := s.locks[clientId]
+	if alreadyExists {
+		logrus.Debugf("Already acquired lock for node %v", clientId)
+		return pb.SemaphoreAccessStatus_LOCKED, nil
+	}
+
+	s.heartBeatMutex.Lock()
+	_, alreadyExists = s.heartbeats[clientId]
+	s.heartbeats[clientId] = time.Now() // update the heartbeat
+	s.heartBeatMutex.Unlock()
+
+	if !alreadyExists {
+		logrus.Debugf("Enqueueing node %v", clientId)
+		s.priorityQ.Enqueue(clientId, priority) // no heartbeat - new node request
+	}
+
+	if lockAcquired := s.tryLock(clientId); lockAcquired {
+		logrus.Infof("Acquired lock for node %v", clientId)
+		updateDone = true
+		return pb.SemaphoreAccessStatus_LOCKED, nil
+	}
+	return pb.SemaphoreAccessStatus_QUEUED, nil
+}
+
+func (s *semaphorePriorityQueue) release(clientId string, lockId uint32) {
+	delete(s.locks, clientId)
+	s.availableLocks = append(s.availableLocks, lockId)
+}
+
+func (s *semaphorePriorityQueue) ReleaseLock(clientId string) error {
+	logrus.Infof("Received ReleaseLock request for node %v - acquiring mutex", clientId)
+	s.mutex.Lock()
+	updateDone := false
+	defer func() {
+		s.mutex.Unlock()
+		if updateDone {
+			<-s.configMapUpdateDone
+		}
+	}()
+	logrus.Infof("Received ReleaseLock request for node %v", clientId)
+
+	// release the lock
+	lockId, ok := s.locks[clientId]
+	if !ok {
+		logrus.Warnf("Did NOT find a lock for the node %v!", clientId)
+		return nil
+	}
+
+	// update internal structures
+	s.heartBeatMutex.Lock()
+	s.heartBeatMutex.Unlock()
+	delete(s.heartbeats, clientId)
+
+	s.release(clientId, lockId)
+	updateDone = true
+
+	return nil
+}
+
+func (s *semaphorePriorityQueue) KeepAlive(clientId string) {
+	s.heartBeatMutex.Lock()
+	defer s.heartBeatMutex.Unlock()
+	logrus.Debugf("Received KeepAlive request for node %v", clientId)
+	s.heartbeats[clientId] = time.Now()
+}
+
+func (s *semaphorePriorityQueue) fetchAvailableLock() uint32 {
+	if len(s.availableLocks) == 0 {
+		return NULL_LOCK
+	}
+	lockId := s.availableLocks[0]
+	s.availableLocks = s.availableLocks[1:]
+	return lockId
+}
+
+func (s *semaphorePriorityQueue) tryLock(clientId string) bool {
+	// check if the node is at the front of the queue
+	nextResouceId, priority := s.priorityQ.Front()
+	if nextResouceId == "" {
+		panic("Queue is empty")
```

Comment:
Can we do something else other than panic here? I would suggest handling this situation